diag/diagnostics_test.go

```diff
@@ -518,3 +519,123 @@ func TestDiagnosticsHasError(t *testing.T) {
 		})
 	}
 }
+
+func TestDiagnosticsErrorsCount(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		diags    diag.Diagnostics
+		expected int
+	}
+	tests := map[string]testCase{
+		"errors": {
+			diags: diag.Diagnostics{
+				diag.NewErrorDiagnostic("Error Summary", "Error detail."),
+				diag.NewWarningDiagnostic("Warning Summary", "Warning detail."),
+			},
+			expected: 1,
+		},
+	}
+
+	for name, test := range tests {
+		name, test := name, test
+		t.Run(name, func(t *testing.T) {
+			got := test.diags.ErrorsCount()
+
+			if diff := cmp.Diff(test.expected, got); diff != "" {
+				t.Fatalf("expected: %q, got: %q", test.expected, got)
+			}
+		})
+	}
+}
+
+func TestDiagnosticsWarningsCount(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		diags    diag.Diagnostics
+		expected int
+	}
+	tests := map[string]testCase{
+		"errors": {
+			diags: diag.Diagnostics{
+				diag.NewErrorDiagnostic("Error Summary", "Error detail."),
+				diag.NewWarningDiagnostic("Warning Summary", "Warning detail."),
+			},
+			expected: 1,
+		},
+	}
+
+	for name, test := range tests {
+		name, test := name, test
+		t.Run(name, func(t *testing.T) {
+			got := test.diags.WarningsCount()
+
+			if diff := cmp.Diff(test.expected, got); diff != "" {
+				t.Fatalf("expected: %q, got: %q", test.expected, got)
+			}
+		})
+	}
+}
+
+func TestDiagnosticsErrors(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		diags    diag.Diagnostics
+		expected diag.Diagnostics
+	}
+	tests := map[string]testCase{
+		"errors": {
+			diags: diag.Diagnostics{
+				diag.NewErrorDiagnostic("Error Summary", "Error detail."),
+				diag.NewWarningDiagnostic("Warning Summary", "Warning detail."),
+			},
+			expected: diag.Diagnostics{
+				diag.NewErrorDiagnostic("Error Summary", "Error detail."),
+			},
+		},
+	}
+
+	for name, test := range tests {
+		name, test := name, test
+		t.Run(name, func(t *testing.T) {
+			got := test.diags.Errors()
+
+			if diff := cmp.Diff(test.expected, got); diff != "" {
+				t.Fatalf("expected: %q, got: %q", test.expected, got)
+			}
+		})
+	}
+}
+
+func TestDiagnosticsWarnings(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		diags    diag.Diagnostics
+		expected diag.Diagnostics
+	}
+	tests := map[string]testCase{
+		"errors": {
```

Comment:
Nit: Might be good to check what happens when:

- There is a `nil` `diag.Diagnostics`, it should return nil or empty `diag.Diagnostics`
- There is an empty `diag.Diagnostics`, it should return empty `diag.Diagnostics`
- There are only error diagnostics, it should return empty `diag.Diagnostics`

Reply:
Added additional test coverage.